test/infrastructure/docker/internal/controllers/dockermachine_controller.go

```diff
@@ -310,7 +310,7 @@ func (r *DockerMachineReconciler) reconcileNormal(ctx context.Context, cluster *
 	// Usually a cloud provider will do this, but there is no docker-cloud provider.
 	// Requeue if there is an error, as this is likely momentary load balancer
 	// state changes during control plane provisioning.
-	if err := externalMachine.SetNodeProviderID(ctx); err != nil {
+	if err := externalMachine.SetNodeProviderID(ctx, r.Client); err != nil {
```

Comment:
I thin that's the client to the management cluster, but we will see it in the tests.

Reply:
Yup it is.

Looks like we currently don't access the apiserver of the workload cluster in the CAPD controllers.

This can be done by setting up the ClusterCacheTracker like this: https://github.com/kubernetes-sigs/cluster-api/blob/abb8f5f55aaf12dc0a9572fc4b4d3570fc8bc528/main.go#L295-L314

Then the `tracker` can be passed into the controllers where we need it and a workload cluster client  can be retrieved via: `tracker.GetClient`.